README.txt
=======
stn3270
=======

Super TN3270 is a thin chocolate layer over py3270 cake.


CHANGES.txt
v0.1.0, 2012-04-13 -- Initial release.


stn3270/test/test.py
# -*- coding: utf-8 -*-

################################################################################
# Super TN3270 - stn3270 - is a thin layer on top of py3270 that 
# enables a simpler navigation over virtual 3270 connections.
# Copyright (C) 2012 Alexandre Drummond
# 
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
# 
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
# 
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
################################################################################

import unittest
from flexmock import flexmock



stn3270/test/test_util.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

################################################################################
# Super TN3270 - stn3270 - is a thin layer on top of py3270 that 
# enables a simpler navigation over virtual 3270 connections.
# Copyright (C) 2012 Alexandre Drummond
# 
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
# 
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
# 
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
################################################################################

from ConfigParser import SafeConfigParser
from flexmock import flexmock
import json
import logging
import re
from navigator import *
from field import *

def config_parser(filename):
    config = SafeConfigParser()
    config.read(filename)
    config.filename = filename
    return config

class JsonScreenException(Exception):
    pass

class JsonScreen:
    """Stores and retrieves Screens from json files.
    
    :param name: unique screen name
    :type name: string
    """
    def __init__(self, name):
        if re.search(r'\W', name):
            raise JsonScreenException(('Nome inválido para a tela "%s": nome só pode'+
                                       ' ter caracteres alfanuméricos e _.') % name)
        self.name = name
    
    def load(self, config, emulator_mock):
        """Loads screen from file
        
        :param config: screens config file as a configuration parser
        :param emulator_mock: Saul Tigh 3270 Emulator mock used to retrieve
            a screen from it
        :type config: SafeConfigParser
        :type emulator_mock: flexmock(Emulator)
        :return: modified emulator_mock prepared to return screen, buffer and status
        """
        if config.has_section(self.name):
            self.hex = json.loads(config.get(self.name, 'hex'))
            self.ascii = json.loads(config.get(self.name, 'ascii'))
            self.status = json.loads(config.get(self.name, 'status'))
        else:
            raise Exception('Screen "%s" not found.' % self.name)
        emulator_mock.should_receive('exec_command').with_args('Ascii').and_return(self.ascii)
        emulator_mock.should_receive('exec_command').with_args('ReadBuffer(Ascii)').and_return(self.hex)
        emulator_mock.status = self.status
        return emulator_mock
    
    def store(self, config, emulator):
        """Stores a screen to a file
        
        :param config: configuration to store screen to
        :param emulator: Saul Tigh 3270 Emulator used to retrieve
            a screen from it
        :type emulator: Emulator
        :type config: SafeConfigParser
        """
        self.ascii = emulator.exec_command('Ascii').data
        self.hex = emulator.exec_command('ReadBuffer(Ascii)').data
        self.status = emulator.status.__dict__
        
        if not config.has_section(self.name):
            config.add_section(self.name)
        config.set(self.name, 'hex', json.dumps(self.hex))
        config.set(self.name, 'ascii', json.dumps(self.ascii))
        config.set(self.name, 'status', json.dumps(self.status))
        with open(config.filename, 'wb') as configfile:
            config.write(configfile)
        return self

if __name__ == "__main__":
    import optparse
    
    parser = optparse.OptionParser()
    parser.add_option("-z", "--host", 
                      action="store", type="string", dest="host", default="10.3.10.3")
    parser.add_option("-c", "--capture-file", 
                      action="store", type="string", dest="capturefile", default="capture.3270session")
    parser.add_option("-v", "--verbose",
                      action="store_true", dest="debug", default=False)
    (options, args) = parser.parse_args()
    
    logging.basicConfig(format='%(levelname)-8s %(message)s',
                        level=(logging.DEBUG if options.debug else logging.INFO))
    config = config_parser(options.capturefile)
    em = tigh3270.Emulator(True)
    em.connect(options.host)
    
    print "Pressione Ctrl+C para terminar.\n"

    while True:
        try:
            screen_name = raw_input('Nome da tela a capturar: ').strip()
            if not screen_name:
                continue
        except KeyboardInterrupt:
            break
        
        if config.has_section(screen_name):
            try:
                confirmation = raw_input("Nome da tela já existe. Confirma sobrescrita?[y/N] ").strip().upper()
                if confirmation in ['N','']:
                    continue
            except KeyboardInterrupt:
                break
        try:
            screen = JsonScreen(screen_name)
            screen.store(config, em)
        except JsonScreenException, e:
            logging.error(str(e))
            

stn3270/field.py
# -*- coding: utf-8 -*-
"""
******************
Super TN3270 Field
******************

Super TN3270 Field - stn3270.field - implements the field
manipulation on a 3270 virtual screen: read, write, verify
and find fields. 
"""

class Field:
    """It's a representation of a 3270 field, with a *start of field* sequence, 
    its position (*row* and *column*), raw *text* and its ASCII *data* representation
    
    :param start_of_field: a 3270 SF sequence
    :param row: starting row of the field
    :param col: starting column of the field
    :param text: raw text of the field
    :param filler: ASCII character used to fill empty editable field
    :type start_of_field: string
    :type row: int
    :type col: int
    :type text: string
    :type filler: string
    """ 
    def __init__(self, start_of_field, row=None, col=None, text="", filler="_"):
        self.filler = filler
        self.start_of_field = self._SF(start_of_field)
        self.row = row
        self.col = col
        self.set_text(text)
        self.is_visible = ("c0=cd" not in start_of_field)
        self.is_editable = False
        for sf in self.start_of_field:
            self.is_editable = self.is_editable or sf in ("c0=c1","c0=cd")

    def set_text(self, text):
        """Sets the text of the field and the filtered data (text without filler characters)
        
        :param text: raw text of field
        :type text: string
        """
        self.text = text
        self.length = len(text)
        self.data = text.replace(self.filler," ").rstrip()
        return self.data

    def _SF(self, char):
        return char[3:-1].split(',')


stn3270/__init__.py
# -*- coding: utf-8 -*-
"""
************
Super TN3270
************

Super TN3270 - stn3270 - is a thin layer on top of py3270 that 
enables a simpler navigation over virtual 3270 connections and easy
form filling and reading. 
"""

################################################################################
# Super TN3270 - stn3270 - is a thin layer on top of py3270 that 
# enables a simpler navigation over virtual 3270 connections.
# Copyright (C) 2012 Alexandre Drummond
# 
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
# 
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
# 
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
################################################################################

import logging
from navigator import *
from field import *

__author__ = "Alexandre A. Drummond Barroso"
__copyright__ = "Copyright 2012, Alexandre Drummond"
__credits__ = ["Alexandre Drummond"]
__license__ = "LGPL v2.1"
__version__ = "0.1"
__maintainer__ = "Alexandre Drummond"
__email__ = "alexandre.drummond@gmail.com"
__status__ = "Prototype" # Unfortunately not "Development", or "Production" yet

__all__ = []


stn3270/navigator.py
# -*- coding: utf-8 -*-

################################################################################
# Super TN3270 - stn3270 - is a thin layer on top of py3270 that 
# enables a simpler navigation over virtual 3270 connections.
# Copyright (C) 2012 Alexandre Drummond
# 
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
# 
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
# 
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
################################################################################

"""
**********************
Super TN3270 Navigator
**********************

Super TN3270 Navigator - stn3270.navigator - implements the core
funcitonality of stn3270, which is retrieving data from and
posting data to a 3270 terminal emulator. 
"""

from py3270 import EmulatorBase
import re
from field import *

class Emulator(EmulatorBase):
    """Represents a virtual 3270 terminal and its operations.
    
    Create an Emulator with:
    
    :param visible: Defines if the virtual 3270 terminal must be visible or run in background.
    :type visible: Boolean
    
    .. IMPORTANT::
       It needs x3270 and s3270 command line tools installed to work.
    """
    x3270_executable = '/usr/bin/x3270'
    s3270_executable = '/usr/bin/s3270'

class Navigator:
    """Automates the navigation between screens, fills text fields, get
    field contents from screen.
    
    To create a Navigator you need:
    
    :param emulator: an Emulator instance
    :param host: host IP or name
    :param label_pattern: a regular expression to identify a label from a
        text field. The label field must be identified by LABELNEXT*id* when
        the label comes after the field or identified by LABELPREV*id* when
        the label comes before the field. Use @ character to identify a
        visible field or _ to identify an invisible one.
    :type emulator: Emulator
    :type host: string
    :type label_pattern: string
    """    
    def __init__(self, emulator, host='10.3.10.3', 
                 label_pattern=r'(\(?\s*(_|@)\s*\)?\s+(?P<LABELNEXT1>.*))|((?P<LABELPREV2>.*?)\s*(:|-->|==>)@)'):
        self.emulator = emulator
        if not self.emulator.status.connection_state:
            self.emulator.connect(host)
        self.rows = int(self.emulator.status.row_number)
        self.cols = int(self.emulator.status.col_number)
        self.label_pattern = label_pattern

    def send(self, command):
        self.emulator.exec_command(command)
        self.emulator.wait_for_field()
        self.screen = self.emulator.exec_command('Ascii').data
        self.buffer = self.emulator.exec_command('ReadBuffer(Ascii)').data
        fields = self._process_fields(self.buffer)
        self.screen_map = self._process_labels(fields, self.label_pattern)
        ## retornar array de campos editáveis com o filter

    def _process_fields(self, input_buffer):
        buffer = [line.split() for line in input_buffer]
        fields = []
        field = None
        for row in range(self.rows):
            for col in range(self.cols):
                if self._is_start_of_field(buffer[row][col]):
                    if field is not None:
                        field.set_text(self.text_from_screen(field.row, field.col, *self._end_of_field_position(row,col)))
                        fields.append(field)
                    field = Field(buffer[row][col], *self._start_of_field_position(row,col))
        return fields

    def _is_start_of_field(self, char):
        return char[:2] == 'SF'

    def _start_of_field_position(self, sf_row, sf_col):
        if sf_col == self.cols - 1:
            result_row = sf_row + 1
            result_col = 0
        else:
            result_row = sf_row
            result_col = sf_col + 1
        return result_row, result_col

    def _end_of_field_position(self, sf_row, sf_col):
        if sf_col == 0:
            result_row = sf_row - 1
            result_col = self.cols - 1
        else:
            result_row = sf_row
            result_col = sf_col - 1
        return result_row, result_col
    
    def text_from_screen(self, start_row, start_col, end_row, end_col):
        if end_col < self.cols and end_row < self.rows and \
                (start_row < end_row or (start_row == end_row and start_col <= end_col)):
            row = start_row
            col = start_col
            text = ""
            while row <= end_row:
                if row == end_row:
                    text += self.screen[row][col:end_col+1]
                else:
                    text += self.screen[row][col:self.cols]
                col = 0
                row += 1
            return text
        else:
            return None

    def _process_labels(self, fields, label_pattern):
        label_re = re.compile(label_pattern)
        num_fields = len(fields)
        re_label_text = re.compile(r'(\w+(\s+\w+)*)')
        for cur_field in range(num_fields):
            text = self._format_fields_for_re(fields, cur_field)
            search_result = label_re.search(text).groupdict()
            for result in search_result:
                if search_result[result] is not None:
                    label_field = cur_field + 1 if "NEXT" in result else cur_field - 1
                    fields[cur_field].label = re_label_text.search(fields[label_field].data).groups()[0]
                    
    def _format_fields_for_re(self, fields, cur_field):
        num_fields = len(fields)
        if cur_field > 0 and fields[cur_field-1].row == fields[cur_field].row:
            previous_field = self._format_for_label_match(fields[cur_field-1])
        else:
            previous_field = ""
        if cur_field < num_fields - 1 and fields[cur_field+1].row == fields[cur_field].row:
            next_field = self._format_for_label_match(fields[cur_field+1])
        else:
            next_field = ""
        field_type_char = "@" if fields[cur_field].is_visible else "_"
        return previous_field + field_type_char + next_field
        
    def _format_for_label_match(self, str):
        return str.strip().replace("@","ø").replace("_","ø")
